File: legacy/tools/motioncheck.py
#!/usr/bin/env python3
"""Everything that animates must stop when the system asks it to.

Ten animations sat outside the prefers-reduced-motion block — the chart draw,
the score bar fill, the stability strips, the splash, the reading-confirmation
flourishes. A partial block is arguably worse than none: the motion that
remains is precisely the motion nobody thought to check, and someone who has
asked their device not to animate should not have to argue with it.

Nothing else inspected this. csscheck confirms every class has a rule and every
rule is used; it has no opinion about whether a rule moves.
"""
import re
import sys

path = sys.argv[1] if len(sys.argv) > 1 else "build/reef-console.html"
html = open(path, encoding="utf-8").read()

# Classes whose rule declares an animation.
animated = {}
for m in re.finditer(r"\.([\w-]+)\s*\{([^}]*)\}", html):
    cls, body = m.group(1), m.group(2)
    if re.search(r"animation:\s*\w", body):
        found = re.search(r"animation:\s*([\w-]+)", body)
        animated[cls] = found.group(1) if found else "?"

# Classes named inside a reduced-motion block.
guarded = set()
for block in re.findall(r"@media \(prefers-reduced-motion[^{]*\{([\s\S]*?)\n\s*\}\s*\n", html):
    guarded |= set(re.findall(r"\.([\w-]+)", block))

if not animated:
    print("FAIL motioncheck: found no animations at all — the parser is wrong")
    sys.exit(1)
if not guarded:
    print("FAIL motioncheck: no prefers-reduced-motion block")
    sys.exit(1)

missing = sorted(c for c in animated if c not in guarded)
if missing:
    print("FAIL animations that ignore prefers-reduced-motion:")
    for cls in missing:
        print(f"    .{cls} ({animated[cls]})")
    sys.exit(1)
print(f"OK   every animation respects reduced motion ({len(animated)} checked)")

File: scripts/verify/motioncheck.mjs
#!/usr/bin/env node
/* Ported from legacy/tools/motioncheck.py.
 *
 * Everything that animates must stop when the system asks it to. Ten
 * animations once sat outside the prefers-reduced-motion block; a partial
 * block is arguably worse than none, since the motion left over is precisely
 * what nobody thought to check.
 *
 * Change from the monolith: reads src/styles/*.css directly instead of a
 * built HTML file's embedded stylesheet. There is no single inlined
 * stylesheet under Tailwind + PostCSS to read post-build, and source is
 * strictly more useful here — no build step needed before this can run.
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT } from './util.mjs';

const cssDir = path.join(ROOT, 'src', 'styles');
const cssFiles = fs.readdirSync(cssDir).filter((f) => f.endsWith('.css'));
const css = cssFiles.map((f) => fs.readFileSync(path.join(cssDir, f), 'utf8')).join('\n');

const animated = {};
for (const m of css.matchAll(/\.([\w-]+)\s*\{([^}]*)\}/g)) {
  const [, cls, body] = m;
  if (/animation:\s*\w/.test(body)) {
    const found = body.match(/animation:\s*([\w-]+)/);
    animated[cls] = found ? found[1] : '?';
  }
}

const guarded = new Set();
for (const block of css.matchAll(/@media \(prefers-reduced-motion[^{]*\{([\s\S]*?)\n\s*\}\s*\n/g)) {
  for (const c of block[1].matchAll(/\.([\w-]+)/g)) guarded.add(c[1]);
}

if (!Object.keys(animated).length) {
  console.log('FAIL motioncheck: found no animations at all — the parser is wrong');
  process.exit(1);
}
if (!guarded.size) {
  console.log('FAIL motioncheck: no prefers-reduced-motion block');
  process.exit(1);
}

const missing = Object.keys(animated).filter((c) => !guarded.has(c)).sort();
if (missing.length) {
  console.log('FAIL animations that ignore prefers-reduced-motion:');
  for (const cls of missing) console.log(`    .${cls} (${animated[cls]})`);
  process.exit(1);
}
console.log(`OK   every animation respects reduced motion (${Object.keys(animated).length} checked, ${cssFiles.length} CSS files)`);
